File: src/lib/database/models/Favorite.ts
import mongoose, { type Document, type Model } from 'mongoose';

export type FavoriteItemType = 'guide' | 'food' | 'hotel' | 'route';

export interface IFavorite extends Document {
  userId: mongoose.Types.ObjectId;
  itemId: string;
  itemType: FavoriteItemType;
  createdAt: Date;
  updatedAt: Date;
}

const favoriteSchema = new mongoose.Schema<IFavorite>(
  {
    userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    itemId: { type: String, required: true },
    itemType: {
      type: String,
      enum: ['guide', 'food', 'hotel', 'route'],
      required: true,
    },
  },
  { timestamps: true }
);

favoriteSchema.index({ userId: 1, itemId: 1 }, { unique: true });
favoriteSchema.index({ itemId: 1, itemType: 1 });

const Favorite: Model<IFavorite> =
  (mongoose.models.Favorite as Model<IFavorite>) ||
  mongoose.model<IFavorite>('Favorite', favoriteSchema);

export default Favorite;

File: backend/internal/model/favorite.go
package model

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type FavoriteItemType string

const (
	FavGuide FavoriteItemType = "guide"
	FavFood  FavoriteItemType = "food"
	FavHotel FavoriteItemType = "hotel"
	FavRoute FavoriteItemType = "route"
)

type Favorite struct {
	ID        primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	UserID    primitive.ObjectID `bson:"userId" json:"userId"`
	ItemID    string             `bson:"itemId" json:"itemId"`
	ItemType  FavoriteItemType   `bson:"itemType" json:"itemType"`
	CreatedAt time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt time.Time          `bson:"updatedAt" json:"updatedAt"`
}
